File: Android/app/src/main/java/com/sans347/app/ui/screens/TablesScreen.kt
package com.sans347.app.ui.screens

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Description
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.foundation.horizontalScroll
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.sans347.app.ui.theme.SansColors

@Composable
fun TablesScreen(modifier: Modifier = Modifier) {
    val scrollState = rememberScrollState()
    BoxWithConstraints(
        modifier = modifier
            .fillMaxSize()
            .background(SansColors.DarkBlueGray),
    ) {
        Box(Modifier.fillMaxSize()) {
            Column(
                modifier = Modifier
                    .widthIn(max = 896.dp)
                    .fillMaxWidth()
                    .align(Alignment.TopCenter)
                    .verticalScroll(scrollState)
                    .padding(start = 16.dp, end = 16.dp + 10.dp, top = 16.dp, bottom = 100.dp),
            ) {
                TablesContent()
            }
            ScrollbarTrack(
                scrollState = scrollState,
                modifier = Modifier
                    .align(Alignment.TopEnd)
                    .width(6.dp)
                    .fillMaxHeight()
                    .padding(top = 16.dp, bottom = 100.dp, end = 4.dp),
            )
        }
    }
}

@Composable
private fun ScrollbarTrack(
    scrollState: androidx.compose.foundation.ScrollState,
    modifier: Modifier = Modifier,
) {
    val max = scrollState.maxValue
    Canvas(modifier = modifier) {
        val trackW = size.width
        val trackH = size.height
        drawRoundRect(
            color = androidx.compose.ui.graphics.Color(0xFF1F2937),
            topLeft = Offset.Zero,
            size = Size(trackW, trackH),
            cornerRadius = CornerRadius(trackW / 2, trackW / 2),
        )
        if (max > 0) {
            val thumbH = (trackH * trackH / (trackH + max)).coerceIn(24f, trackH * 0.9f)
            val travel = (trackH - thumbH).coerceAtLeast(0f)
            val progress = scrollState.value / max.toFloat()
            val thumbY = travel * progress
            drawRoundRect(
                color = androidx.compose.ui.graphics.Color(0xFF4B5563),
                topLeft = Offset(0f, thumbY),
                size = Size(trackW, thumbH),
                cornerRadius = CornerRadius(3.dp.toPx(), 3.dp.toPx()),
            )
        }
    }
}

@Composable
private fun TablesContent() {
        Column(verticalArrangement = Arrangement.spacedBy(24.dp)) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .background(SansColors.Gray700, RoundedCornerShape(12.dp))
                .border(1.dp, SansColors.Gray600, RoundedCornerShape(12.dp)),
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .border(1.dp, SansColors.Gray600, RoundedCornerShape(topStart = 12.dp, topEnd = 12.dp))
                    .padding(16.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(Icons.Default.Description, null, tint = SansColors.White, modifier = Modifier.padding(end = 12.dp))
                Column {
                    Text("Reference Tables", color = SansColors.White, fontWeight = FontWeight.Bold, fontSize = 20.sp)
                    Text("SANS 347 - 2024 3rd Edition", color = SansColors.Gray400, fontSize = 14.sp)
                }
            }
        }

        Table1Card()
        Table2Card()
        Table3Card()
    }
}

@Composable
private fun Table1Card() {
    DarkTableCard(title = "Table 1 — Categorization Figures") {
        Column {
            Table1Body()
            Column(Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(4.dp)) {
                NoteLine("NOTE:", "For two-phase flow, the equipment should be categorized to the higher risk.")
                NoteLine("a", "Transportable gas container and their safety and pressure accessories shall be assessed using table 3.")
                NoteLine("b", "No pockets of gas may form above the liquid in the equipment, including steam.")
                NoteLine("c", "Fluid group 1 = dangerous; fluid group 2 = not dangerous.")
            }
        }
    }
}

@Composable
private fun Table1Body() {
    BoxWithConstraints(Modifier.padding(8.dp)) {
        val isWide = maxWidth >= 500.dp
        if (isWide) {
            Table1Wide()
        } else {
            Table1Narrow()
        }
    }
}

@Composable
private fun Table1Narrow() {
    val hScroll = rememberScrollState()
    val c = 34.dp
    val labelW = 100.dp
    Row(Modifier.horizontalScroll(hScroll)) {
        Column(
            modifier = Modifier.background(SansColors.Gray700),
        ) {
            Row(Modifier.background(SansColors.Gray600)) {
                TableCell("Equipment Type", bold = true, header = true, width = labelW, smallText = true)
                TableCell("Pressure Vessels", bold = true, header = true, width = c * 4, center = true, smallHeader = true)
                TableCell("Steam Gen.", bold = true, header = true, width = c, center = true, smallHeader = true)
                TableCell("Piping", bold = true, header = true, width = c * 4, center = true, smallHeader = true)
                TableCell("TGC", bold = true, header = true, width = c, center = true, smallHeader = true)
            }
            Row {
                TableCell("State of Contents", bold = true, width = labelW, smallText = true)
                TableCell("Gas", width = c * 2, center = true)
                TableCell("Liquid", width = c * 2, center = true)
                TableCell("—", width = c, center = true)
                TableCell("Gas", width = c * 2, center = true)
                TableCell("Liquid\u1D47", width = c * 2, center = true)
                TableCell("Gas", width = c, center = true)
            }
            Row {
                TableCell("Fluid Group\u1D9C", bold = true, width = labelW, smallText = true)
                TableCell("1", width = c, center = true)
                TableCell("2", width = c, center = true)
                TableCell("1", width = c, center = true)
                TableCell("2", width = c, center = true)
                TableCell("—", width = c, center = true)
                TableCell("1", width = c, center = true)
                TableCell("2", width = c, center = true)
                TableCell("1", width = c, center = true)
                TableCell("2", width = c, center = true)
                TableCell("1", width = c, center = true)
            }
            Row {
                TableCell("Refer to Figure", bold = true, width = labelW, smallText = true)
                TableCell("1", fig = true, width = c, center = true)
                TableCell("2", fig = true, width = c, center = true)
                TableCell("3", fig = true, width = c, center = true)
                TableCell("4", fig = true, width = c, center = true)
                TableCell("5", fig = true, width = c, center = true)
                TableCell("6", fig = true, width = c, center = true)
                TableCell("7", fig = true, width = c, center = true)
                TableCell("8", fig = true, width = c, center = true)
                TableCell("9", fig = true, width = c, center = true)
                TableCell("a", width = c, center = true)
            }
        }
    }
}

@Composable
private fun Table1Wide() {
    val lw = 1.8f
    val cw = 1f

    Column(Modifier.fillMaxWidth().background(SansColors.Gray700)) {
        Row(Modifier.fillMaxWidth().background(SansColors.Gray600)) {
            WideCell("Equipment Type", Modifier.weight(lw), bold = true, header = true)
            WideCell("Pressure Vessels", Modifier.weight(cw * 4), bold = true, header = true, center = true)
            WideCell("Steam", Modifier.weight(cw), bold = true, header = true, center = true)
            WideCell("Piping", Modifier.weight(cw * 4), bold = true, header = true, center = true)
            WideCell("TGC", Modifier.weight(cw), bold = true, header = true, center = true)
        }
        Row(Modifier.fillMaxWidth()) {
            WideCell("State of Contents", Modifier.weight(lw), bold = true)
            WideCell("Gas", Modifier.weight(cw * 2), center = true)
            WideCell("Liquid", Modifier.weight(cw * 2), center = true)
            WideCell("—", Modifier.weight(cw), center = true)
            WideCell("Gas", Modifier.weight(cw * 2), center = true)
            WideCell("Liquid\u1D47", Modifier.weight(cw * 2), center = true)
            WideCell("Gas", Modifier.weight(cw), center = true)
        }
        Row(Modifier.fillMaxWidth()) {
            WideCell("Fluid Group\u1D9C", Modifier.weight(lw), bold = true)
            WideCell("1", Modifier.weight(cw), center = true)
            WideCell("2", Modifier.weight(cw), center = true)
            WideCell("1", Modifier.weight(cw), center = true)
            WideCell("2", Modifier.weight(cw), center = true)
            WideCell("—", Modifier.weight(cw), center = true)
            WideCell("1", Modifier.weight(cw), center = true)
            WideCell("2", Modifier.weight(cw), center = true)
            WideCell("1", Modifier.weight(cw), center = true)
            WideCell("2", Modifier.weight(cw), center = true)
            WideCell("1", Modifier.weight(cw), center = true)
        }
        Row(Modifier.fillMaxWidth()) {
            WideCell("Refer to Figure", Modifier.weight(lw), bold = true)
            WideCell("1", Modifier.weight(cw), fig = true, center = true)
            WideCell("2", Modifier.weight(cw), fig = true, center = true)
            WideCell("3", Modifier.weight(cw), fig = true, center = true)
            WideCell("4", Modifier.weight(cw), fig = true, center = true)
            WideCell("5", Modifier.weight(cw), fig = true, center = true)
            WideCell("6", Modifier.weight(cw), fig = true, center = true)
            WideCell("7", Modifier.weight(cw), fig = true, center = true)
            WideCell("8", Modifier.weight(cw), fig = true, center = true)
            WideCell("9", Modifier.weight(cw), fig = true, center = true)
            WideCell("a", Modifier.weight(cw), center = true)
        }
    }
}

@Composable
private fun RowScope.WideCell(
    text: String,
    modifier: Modifier = Modifier,
    bold: Boolean = false,
    header: Boolean = false,
    fig: Boolean = false,
    center: Boolean = false,
) {
    Text(
        text,
        modifier = modifier
            .border(0.5.dp, SansColors.Gray600)
            .padding(horizontal = 4.dp, vertical = 6.dp),
        color = when {
            header -> Color.White
            fig -> Color(0xFF3B82F6)
            bold -> SansColors.Gray300
            else -> SansColors.Gray200
        },
        fontWeight = if (bold || fig || header) FontWeight.Bold else FontWeight.Normal,
        fontSize = 11.sp,
        maxLines = 1,
        textAlign = if (center) TextAlign.Center else TextAlign.Start,
    )
}

@Composable
private fun RowScope.TableCell(
    text: String,
    bold: Boolean = false,
    header: Boolean = false,
    fig: Boolean = false,
    width: androidx.compose.ui.unit.Dp,
    center: Boolean = false,
    smallHeader: Boolean = false,
    smallText: Boolean = false,
) {
    Text(
        text,
        modifier = Modifier
            .width(width)
            .border(0.5.dp, SansColors.Gray600)
            .padding(4.dp),
        color = when {
            header -> Color.White
            fig -> Color(0xFF3B82F6)
            bold -> SansColors.Gray300
            else -> SansColors.Gray200
        },
        fontWeight = if (bold || fig || header) FontWeight.Bold else FontWeight.Normal,
        fontSize = when {
            smallHeader -> 8.sp
            smallText -> 9.sp
            else -> 10.sp
        },
        maxLines = if (smallHeader || smallText) 1 else 3,
        textAlign = if (center) TextAlign.Center else TextAlign.Start,
    )
}

@Composable
private fun Table2Card() {
    DarkTableCard(title = "Table 2 — Conformity Assessment Modules for Each Category") {
        Column {
            Column(Modifier.padding(8.dp)) {
                Row(Modifier.fillMaxWidth().background(SansColors.Gray600)) {
                    Text("Hazard Category", Modifier.weight(0.25f).padding(12.dp), color = SansColors.Gray300, fontWeight = FontWeight.SemiBold, fontSize = 13.sp)
                    Text("Manufacturer without Certified Quality System", Modifier.weight(0.375f).padding(12.dp), color = Color.White, fontWeight = FontWeight.SemiBold, fontSize = 13.sp)
                    Text("Manufacturer with Certified Quality System", Modifier.weight(0.375f).padding(12.dp), color = Color.White, fontWeight = FontWeight.SemiBold, fontSize = 13.sp)
                }
                ModuleRow("I", Color(0xFF3B82F6), "A", "A")
                ModuleRow("II", Color(0xFFEAB308), "A2", "A2 or D1 or E1")
                ModuleRow("III", Color(0xFFF97316), "B (design type) + F or\nB (production type) + C2", "H or\nB (production type) + E or\nB (design type) + D")
                ModuleRow("IV", Color(0xFFEF4444), "G or\nB (production type) + F", "H1 or\nB (production type) + D")
            }
            Column(Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(4.dp)) {
                Text("Module Definitions:", color = SansColors.Gray300, fontWeight = FontWeight.SemiBold, fontSize = 12.sp)
                DefLine("A", "internal production control")
                DefLine("A2", "internal production control plus supervised pressure equipment checks at random intervals")
                DefLine("B", "type examination — production type or type examination — design type")
                DefLine("C2", "conformity to type based on internal production control plus supervised pressure equipment checks at random intervals")
                DefLine("D", "conformity to type based on quality assurance of the production process")
                DefLine("D1", "quality assurance of the production process")
                DefLine("E", "conformity to type based on pressure equipment quality assurance")
                DefLine("E1", "product quality assurance for final pressure equipment inspection and testing")
                DefLine("F", "conformity to type based on pressure equipment verification")
                DefLine("G", "conformity based on unit verification")
                DefLine("H", "conformity based on full quality assurance")
                DefLine("H1", "conformity based on full quality assurance plus design examination")
                Spacer(Modifier.height(8.dp))
                NoteLine("NOTE 1:", "For RSA/CI/OHSA marked pressure equipment intended for non-nuclear use, refer to annex B.")
                NoteLine("NOTE 2:", "For RSA/CI/OHSA marked pressure equipment intended for nuclear use, refer to annex C.")
                NoteLine("NOTE 3:", "For non-RSA/CI/OHSA marked pressure equipment intended for nuclear use, refer to annex C.")
            }
        }
    }
}

@Composable
private fun ModuleRow(cat: String, color: Color, without: String, with: String) {
    Row(Modifier.fillMaxWidth().border(1.dp, SansColors.Gray600)) {
        Text(cat, Modifier.width(48.dp).padding(12.dp), color = color, fontWeight = FontWeight.Bold, fontSize = 18.sp)
        Text(without, Modifier.weight(1f).padding(12.dp), color = SansColors.Gray200, fontSize = 14.sp)
        Text(with, Modifier.weight(1f).padding(12.dp), color = SansColors.Gray200, fontSize = 14.sp)
    }
}

@Composable
private fun Table3Card() {
    DarkTableCard(title = "Table 3 — Conformity Assessment Modules for Transportable Gas Containers") {
        Column {
            Text(
                "National legislation requires that all pressure equipment, including transportable gas containers, shall be categorized in accordance with this standard. Transportable gas containers manufactured in accordance with a relevant health and safety standard shall be deemed to comply with the categorization requirements of this standard.",
                color = SansColors.Gray300,
                fontSize = 12.sp,
                modifier = Modifier.padding(16.dp),
            )
            Row(Modifier.fillMaxWidth().background(SansColors.Gray600)) {
                Text("Hazard Category", Modifier.weight(0.4f).padding(12.dp), color = SansColors.Gray300, fontWeight = FontWeight.SemiBold)
                Text("Conformity Assessment Modules a and b", Modifier.weight(0.6f).padding(12.dp), color = Color.White, fontWeight = FontWeight.SemiBold, fontSize = 13.sp)
            }
            Row(Modifier.fillMaxWidth().border(1.dp, SansColors.Gray600)) {
                Text("III", Modifier.weight(0.4f).padding(12.dp), color = Color(0xFFF97316), fontWeight = FontWeight.Bold, fontSize = 18.sp)
                Text("B + F", Modifier.weight(0.6f).padding(12.dp), color = SansColors.Gray200)
            }
            Column(Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(4.dp)) {
                NoteLine("NOTE:", "Table 3 covers test pressures 0 kPa to 300 000 kPa and volume 0,5 L to 3 000 L (water capacity).")
                DefLine("B", "type examination — design type")
                DefLine("F", "conformity to type based on pressure equipment verification")
                Spacer(Modifier.height(8.dp))
                NoteLine("a", "Imported transportable gas containers from the European Union shall comply with the Transportable Pressure Equipment Directive (TPED) 2010/35/EU requirements.")
                NoteLine("b", "Imported transportable gas containers from the United Kingdom shall bear the Rho (π) symbol and the UKCA mark in accordance with the UK Carriage of Dangerous Goods and Use of Transportable Pressure Equipment Regulations 2009.")
            }
        }
    }
}

@Composable
private fun DarkTableCard(title: String, content: @Composable () -> Unit) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(SansColors.Gray700, RoundedCornerShape(12.dp))
            .border(1.dp, SansColors.Gray600, RoundedCornerShape(12.dp)),
    ) {
        Text(
            title,
            modifier = Modifier
                .fillMaxWidth()
                .background(SansColors.Gray600)
                .padding(16.dp),
            color = Color.White,
            fontWeight = FontWeight.Bold,
            fontSize = 18.sp,
        )
        content()
    }
}

@Composable
private fun NoteLine(label: String, body: String) {
    Text(buildString {
        append(label)
        append(" ")
        append(body)
    }, fontSize = 12.sp, color = SansColors.Gray400)
}

@Composable
private fun DefLine(code: String, body: String) {
    Text(buildString {
        append(code)
        append(" = ")
        append(body)
    }, fontSize = 12.sp, color = SansColors.Gray400)
}

File: iOS/SANS347/Views/Screens/TablesScreen.swift
import SwiftUI

/// Tables page, mirroring `TablesScreen.kt` — the dark-themed reference
/// Tables 1-3 from SANS 347:2024. Uses the narrow (horizontally scrollable)
/// Table 1 layout, appropriate for phone widths.
struct TablesScreen: View {
    var body: some View {
        ScrollView {
            VStack(spacing: 24) {
                headerCard
                table1Card
                table2Card
                table3Card
            }
            .maxContentWidth(896)
            .padding(.horizontal, 16)
            .padding(.top, 16)
            .padding(.bottom, 100)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(SansColors.darkBlueGray)
    }

    private var headerCard: some View {
        HStack(spacing: 12) {
            Image(systemName: "doc.text").foregroundColor(.white)
            VStack(alignment: .leading, spacing: 0) {
                Text("Reference Tables").font(.system(size: 20, weight: .bold)).foregroundColor(.white)
                Text("SANS 347 - 2024 3rd Edition").font(.system(size: 14)).foregroundColor(SansColors.gray400)
            }
            Spacer()
        }
        .padding(16)
        .frame(maxWidth: .infinity)
        .background(SansColors.gray700)
        .clipShape(RoundedRectangle(cornerRadius: 12))
        .overlay(RoundedRectangle(cornerRadius: 12).stroke(SansColors.gray600, lineWidth: 1))
    }

    // MARK: - Table 1

    private var table1Card: some View {
        darkCard(title: "Table 1 — Categorization Figures") {
            VStack(alignment: .leading, spacing: 0) {
                ScrollView(.horizontal, showsIndicators: true) {
                    table1Body
                }
                .padding(8)
                VStack(alignment: .leading, spacing: 4) {
                    note("NOTE:", "For two-phase flow, the equipment should be categorized to the higher risk.")
                    note("a", "Transportable gas container and their safety and pressure accessories shall be assessed using table 3.")
                    note("b", "No pockets of gas may form above the liquid in the equipment, including steam.")
                    note("c", "Fluid group 1 = dangerous; fluid group 2 = not dangerous.")
                }
                .padding(16)
            }
        }
    }

    private let cw: CGFloat = 34
    private let labelW: CGFloat = 100

    private var table1Body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 0) {
                cell("Equipment Type", width: labelW, bold: true, header: true, fontSize: 9)
                cell("Pressure Vessels", width: cw * 4, bold: true, header: true, center: true, fontSize: 8)
                cell("Steam Gen.", width: cw, bold: true, header: true, center: true, fontSize: 8)
                cell("Piping", width: cw * 4, bold: true, header: true, center: true, fontSize: 8)
                cell("TGC", width: cw, bold: true, header: true, center: true, fontSize: 8)
            }
            HStack(spacing: 0) {
                cell("State of Contents", width: labelW, bold: true, fontSize: 9)
                cell("Gas", width: cw * 2, center: true)
                cell("Liquid", width: cw * 2, center: true)
                cell("—", width: cw, center: true)
                cell("Gas", width: cw * 2, center: true)
                cell("Liquid\u{1D47}", width: cw * 2, center: true)
                cell("Gas", width: cw, center: true)
            }
            HStack(spacing: 0) {
                cell("Fluid Group\u{1D9C}", width: labelW, bold: true, fontSize: 9)
                ForEach(Array(["1", "2", "1", "2", "—", "1", "2", "1", "2", "1"].enumerated()), id: \.offset) { _, value in
                    cell(value, width: cw, center: true)
                }
            }
            HStack(spacing: 0) {
                cell("Refer to Figure", width: labelW, bold: true, fontSize: 9)
                ForEach(1...9, id: \.self) { i in
                    cell("\(i)", width: cw, fig: true, center: true)
                }
                cell("a", width: cw, center: true)
            }
        }
        .background(SansColors.gray700)
    }

    private func cell(_ text: String, width: CGFloat, bold: Bool = false, header: Bool = false, fig: Bool = false, center: Bool = false, fontSize: CGFloat = 10) -> some View {
        let color: Color = header ? .white : (fig ? SansColors.categoryI : (bold ? SansColors.gray300 : SansColors.gray200))
        return Text(text)
            .font(.system(size: fontSize, weight: (bold || fig || header) ? .bold : .regular))
            .foregroundColor(color)
            .multilineTextAlignment(center ? .center : .leading)
            .frame(width: width, alignment: center ? .center : .leading)
            .padding(4)
            .overlay(Rectangle().stroke(SansColors.gray600, lineWidth: 0.5))
    }

    // MARK: - Table 2

    private var table2Card: some View {
        darkCard(title: "Table 2 — Conformity Assessment Modules for Each Category") {
            VStack(alignment: .leading, spacing: 0) {
                VStack(spacing: 0) {
                    HStack(spacing: 0) {
                        Text("Hazard Category").frame(maxWidth: .infinity, alignment: .leading).padding(12)
                            .font(.system(size: 13, weight: .semibold)).foregroundColor(SansColors.gray300)
                        Text("Manufacturer without Certified Quality System").frame(maxWidth: .infinity, alignment: .leading).padding(12)
                            .font(.system(size: 13, weight: .semibold)).foregroundColor(.white)
                        Text("Manufacturer with Certified Quality System").frame(maxWidth: .infinity, alignment: .leading).padding(12)
                            .font(.system(size: 13, weight: .semibold)).foregroundColor(.white)
                    }
                    .background(SansColors.gray600)
                    moduleRow("I", SansColors.categoryI, "A", "A")
                    moduleRow("II", SansColors.categoryII, "A2", "A2 or D1 or E1")
                    moduleRow("III", SansColors.categoryIII, "B (design type) + F or\nB (production type) + C2", "H or\nB (production type) + E or\nB (design type) + D")
                    moduleRow("IV", SansColors.categoryIV, "G or\nB (production type) + F", "H1 or\nB (production type) + D")
                }
                .padding(8)
                VStack(alignment: .leading, spacing: 4) {
                    Text("Module Definitions:").font(.system(size: 12, weight: .semibold)).foregroundColor(SansColors.gray300)
                    ForEach(Self.moduleDefinitions, id: \.0) { code, body in
                        def(code, body)
                    }
                    Spacer().frame(height: 8)
                    ForEach(Self.table2Notes, id: \.0) { label, body in
                        note(label, body)
                    }
                }
                .padding(16)
            }
        }
    }

    private func moduleRow(_ cat: String, _ color: Color, _ without: String, _ with: String) -> some View {
        HStack(alignment: .top, spacing: 0) {
            Text(cat).font(.system(size: 18, weight: .bold)).foregroundColor(color)
                .frame(width: 48, alignment: .leading).padding(12)
            Text(without).font(.system(size: 14)).foregroundColor(SansColors.gray200)
                .frame(maxWidth: .infinity, alignment: .leading).padding(12)
            Text(with).font(.system(size: 14)).foregroundColor(SansColors.gray200)
                .frame(maxWidth: .infinity, alignment: .leading).padding(12)
        }
        .overlay(Rectangle().stroke(SansColors.gray600, lineWidth: 1))
    }

    // MARK: - Table 3

    private var table3Card: some View {
        darkCard(title: "Table 3 — Conformity Assessment Modules for Transportable Gas Containers") {
            VStack(alignment: .leading, spacing: 0) {
                Text("National legislation requires that all pressure equipment, including transportable gas containers, shall be categorized in accordance with this standard. Transportable gas containers manufactured in accordance with a relevant health and safety standard shall be deemed to comply with the categorization requirements of this standard.")
                    .font(.system(size: 12)).foregroundColor(SansColors.gray300).padding(16)
                HStack(spacing: 0) {
                    Text("Hazard Category").frame(maxWidth: .infinity, alignment: .leading).padding(12)
                        .font(.system(size: 14, weight: .semibold)).foregroundColor(SansColors.gray300)
                    Text("Conformity Assessment Modules a and b").frame(maxWidth: .infinity, alignment: .leading).padding(12)
                        .font(.system(size: 13, weight: .semibold)).foregroundColor(.white)
                }
                .background(SansColors.gray600)
                HStack(spacing: 0) {
                    Text("III").font(.system(size: 18, weight: .bold)).foregroundColor(SansColors.categoryIII)
                        .frame(maxWidth: .infinity, alignment: .leading).padding(12)
                    Text("B + F").foregroundColor(SansColors.gray200)
                        .frame(maxWidth: .infinity, alignment: .leading).padding(12)
                }
                .overlay(Rectangle().stroke(SansColors.gray600, lineWidth: 1))
                VStack(alignment: .leading, spacing: 4) {
                    note("NOTE:", "Table 3 covers test pressures 0 kPa to 300 000 kPa and volume 0,5 L to 3 000 L (water capacity).")
                    def("B", "type examination — design type")
                    def("F", "conformity to type based on pressure equipment verification")
                    Spacer().frame(height: 8)
                    note("a", "Imported transportable gas containers from the European Union shall comply with the Transportable Pressure Equipment Directive (TPED) 2010/35/EU requirements.")
                    note("b", "Imported transportable gas containers from the United Kingdom shall bear the Rho (π) symbol and the UKCA mark in accordance with the UK Carriage of Dangerous Goods and Use of Transportable Pressure Equipment Regulations 2009.")
                }
                .padding(16)
            }
        }
    }

    // MARK: - Shared

    private func darkCard<Content: View>(title: String, @ViewBuilder content: () -> Content) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            Text(title)
                .font(.system(size: 18, weight: .bold)).foregroundColor(.white)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(16)
                .background(SansColors.gray600)
            content()
        }
        .frame(maxWidth: .infinity)
        .background(SansColors.gray700)
        .clipShape(RoundedRectangle(cornerRadius: 12))
        .overlay(RoundedRectangle(cornerRadius: 12).stroke(SansColors.gray600, lineWidth: 1))
    }

    private func note(_ label: String, _ body: String) -> some View {
        Text("\(label) \(body)")
            .font(.system(size: 12)).foregroundColor(SansColors.gray400)
            .frame(maxWidth: .infinity, alignment: .leading)
    }

    private func def(_ code: String, _ body: String) -> some View {
        Text("\(code) = \(body)")
            .font(.system(size: 12)).foregroundColor(SansColors.gray400)
            .frame(maxWidth: .infinity, alignment: .leading)
    }

    private static let moduleDefinitions: [(String, String)] = [
        ("A", "internal production control"),
        ("A2", "internal production control plus supervised pressure equipment checks at random intervals"),
        ("B", "type examination — production type or type examination — design type"),
        ("C2", "conformity to type based on internal production control plus supervised pressure equipment checks at random intervals"),
        ("D", "conformity to type based on quality assurance of the production process"),
        ("D1", "quality assurance of the production process"),
        ("E", "conformity to type based on pressure equipment quality assurance"),
        ("E1", "product quality assurance for final pressure equipment inspection and testing"),
        ("F", "conformity to type based on pressure equipment verification"),
        ("G", "conformity based on unit verification"),
        ("H", "conformity based on full quality assurance"),
        ("H1", "conformity based on full quality assurance plus design examination"),
    ]

    private static let table2Notes: [(String, String)] = [
        ("NOTE 1:", "For RSA/CI/OHSA marked pressure equipment intended for non-nuclear use, refer to annex B."),
        ("NOTE 2:", "For RSA/CI/OHSA marked pressure equipment intended for nuclear use, refer to annex C."),
        ("NOTE 3:", "For non-RSA/CI/OHSA marked pressure equipment intended for nuclear use, refer to annex C."),
    ]
}
